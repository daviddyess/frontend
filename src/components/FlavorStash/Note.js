import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import React, { Component, Fragment } from 'react';
import { bindActionCreators } from 'redux';
import unified from 'unified';
import parse from 'remark-parse';
import remark2react from 'remark-react';
import { Form as FinalForm, Field } from 'react-final-form';
import { Button, ButtonGroup, Col, Form, InputGroup } from 'react-bootstrap';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { actions as noteActions } from 'reducers/note';
import { getFlavorNote, isLoading } from 'selectors/note';

export class Note extends Component {
  static propTypes = {
    actions: PropTypes.shape({
      createNote: PropTypes.func.isRequired,
      deleteNote: PropTypes.func.isRequired,
      requestNote: PropTypes.func.isRequired,
      updateNote: PropTypes.func.isRequired
    }),
    flavorId: PropTypes.string.isRequired,
    loading: PropTypes.any.isRequired,
    note: PropTypes.object.isRequired,
    userId: PropTypes.string.isRequired
  };

  constructor(props) {
    super(props);

    this.state = {
      editingNote: false,
      updatedNote: null
    };

    this.handleNoteSubmit = this.handleNoteSubmit.bind(this);
  }

  componentDidMount() {
    const { actions, flavorId } = this.props;

    actions.requestNote({ flavorId });
  }

  addNoteIcon(id) {
    const { editingNote } = this.state;

    return (
      <Fragment>
        {!editingNote && (
          <Button
            className="button-animation"
            size="sm"
            onClick={() => this.handleNoteEditor(id)}
          >
            <FontAwesomeIcon icon="plus" size="sm" title="Add Flavor Note" />
          </Button>
        )}
      </Fragment>
    );
  }

  editNoteIcon(id) {
    const { editingNote } = this.state;

    return (
      <Fragment>
        {!editingNote && (
          <Button
            className="button-animation"
            size="sm"
            onClick={() => this.handleNoteEditor(id)}
          >
            <FontAwesomeIcon icon="pen" size="sm" title="Edit Flavor Note" />
          </Button>
        )}
      </Fragment>
    );
  }

  handleNoteEditor() {
    this.setState({ editingNote: !this.state.editingNote });
  }

  handleNoteSubmit(values) {
    const { actions } = this.props;
    const { note, update } = values;

    if (!note) {
      actions.deleteNote(values);
    } else if (update === true) {
      actions.updateNote(values);
    } else {
      actions.createNote(values);
    }
    this.setState({ updatedNote: note || '' });

    this.handleNoteEditor();
  }

  noteEditor() {
    const { flavorId, note, userId } = this.props;
    const { updatedNote } = this.state;
    const update = Boolean(note?.[flavorId]?.note);

    return (
      <FinalForm
        onSubmit={this.handleNoteSubmit}
        initialValues={{
          flavorId,
          note: updatedNote || note?.[flavorId]?.note,
          update,
          userId
        }}
        render={({ handleSubmit, submitting, values }) => (
          <Form noValidate onSubmit={handleSubmit}>
            <Form.Row>
              <Field name="note">
                {({ input }) => (
                  <Form.Group as={Col}>
                    <InputGroup>
                      <Form.Control
                        {...input}
                        as="textarea"
                        placeholder="Flavor Note"
                      />
                    </InputGroup>
                  </Form.Group>
                )}
              </Field>
            </Form.Row>
            {values?.note ? (
              <Form.Row>
                <h4>Preview:</h4>
                <Col md="12">
                  {
                    /* eslint-disable-next-line no-sync */
                    unified()
                      .use(parse)
                      .use(remark2react)
                      .processSync(values.note).result
                  }
                </Col>
              </Form.Row>
            ) : null}
            <Field name="userId">
              {({ input }) => <Form.Control {...input} type="hidden" />}
            </Field>
            <Field name="flavorId">
              {({ input }) => <Form.Control {...input} type="hidden" />}
            </Field>
            <Field name="update">
              {({ input }) => <Form.Control {...input} type="hidden" />}
            </Field>
            <Form.Row>
              <Form.Group>
                <ButtonGroup>
                  <Button
                    className="button-animation"
                    variant="primary"
                    type="submit"
                    disabled={submitting}
                  >
                    <FontAwesomeIcon icon="save" size="sm" />
                    &nbsp;<span>Save</span>
                  </Button>
                  <Button
                    onClick={e => this.handleNoteEditor(false, e)}
                    className="button-animation button--cancel"
                  >
                    <span>Cancel</span>
                  </Button>
                </ButtonGroup>
              </Form.Group>
            </Form.Row>
          </Form>
        )}
      />
    );
  }

  render() {
    const { note, flavorId, loading } = this.props;
    const { editingNote } = this.state;
    const updatedNote = this.state.updatedNote;

    if (loading !== flavorId) {
      return (
        <Fragment>
          <h3>
            {note[flavorId] || updatedNote
              ? this.editNoteIcon(flavorId)
              : this.addNoteIcon(flavorId)}{' '}
            Flavor Notes
          </h3>
          {!editingNote && (note[flavorId] || updatedNote) ? (
            <Fragment>
              {
                /* eslint-disable-next-line no-sync */
                unified()
                  .use(parse)
                  .use(remark2react)
                  .processSync(updatedNote || note[flavorId].note).result
              }
            </Fragment>
          ) : null}
          {editingNote ? this.noteEditor() : null}
        </Fragment>
      );
    } else {
      return <Fragment>Loading...</Fragment>;
    }
  }
}

export const mapStateToProps = state => ({
  loading: isLoading(state),
  note: getFlavorNote(state)
});

export const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators(
    {
      ...noteActions
    },
    dispatch
  )
});

export default connect(mapStateToProps, mapDispatchToProps)(Note);
