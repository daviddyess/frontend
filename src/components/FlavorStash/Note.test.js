import React from 'react';
import renderer from 'react-test-renderer';
import { withMemoryRouter } from 'utils/testing';

import { Note, mapStateToProps, mapDispatchToProps } from './Note';

jest.mock('react-redux', () => {
  const { connect: rawConnect } = jest.requireActual('react-redux');

  return {
    connect: jest.fn(rawConnect)
  };
});

const RoutedNote = withMemoryRouter(Note);

describe('<Note />', () => {
  const actions = {
    createNote: jest.fn(),
    deleteNote: jest.fn(),
    requestNote: jest.fn(),
    updateNote: jest.fn()
  };
  const props = {
    actions,
    flavorId: '23',
    loading: false,
    note: {
      23: {
        text: "It's dericious!"
      }
    },
    userId: '17'
  };

  const component = renderer.create(<RoutedNote {...props} />);

  const { instance } = component.root.findByType(Note);

  it('can render', () => {
    const tree = renderer.create(<RoutedNote {...props} />).toJSON();

    expect(tree).toMatchSnapshot();
    expect(actions.requestNote).toHaveBeenCalled();
  });

  it('can render with an empty note', () => {
    const tree = renderer.create(<RoutedNote {...props} note={{}} />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('calls requestStash on mount', () => {
    renderer.create(<RoutedNote {...props} stashLoaded={false} />);

    expect(actions.requestNote).toHaveBeenCalled();
  });

  it('can mapStateToProps', () => {
    const loading = false;
    const note = {};
    const state = {
      note: {
        collection: {},
        loading
      }
    };
    const expected = {
      note,
      loading
    };

    expect(mapStateToProps(state)).toEqual(expected);
  });

  it('can mapDispatchToProps', () => {
    const dispatch = jest.fn();
    const expected = {
      actions: expect.objectContaining({
        requestNote: expect.any(Function)
      })
    };

    expect(mapDispatchToProps(dispatch)).toEqual(expected);
  });

  it('can get note edit icon', () => {
    expect(instance).toBeDefined();
    instance.state.editingNote = false;
    expect(instance.editNoteIcon(2)).toMatchSnapshot();
  });

  it('can get add note icon', () => {
    expect(instance).toBeDefined();
    instance.state.editingNote = false;
    expect(instance.editNoteIcon(1)).toMatchSnapshot();
  });

  it('can handle note editor', () => {
    expect(instance).toBeDefined();
    instance.handleNoteEditor();
    expect(instance.state.editingNote).toEqual(true);
  });

  it('can handleNoteSubmit create', () => {
    expect(instance).toBeDefined();
    const values = { flavorId: 2, note: 'Yummy', update: false };

    instance.handleNoteEditor = jest.fn();
    instance.handleNoteSubmit(values);
    expect(actions.createNote).toBeCalledWith(values);
    expect(instance.state.updatedNote).toEqual('Yummy');
    expect(instance.handleNoteEditor).toHaveBeenCalled();
  });

  it('can handleNoteSubmit update', () => {
    expect(instance).toBeDefined();
    const values = { flavorId: 2, note: 'Yummy!', update: true };

    instance.handleNoteSubmit(values);
    expect(actions.updateNote).toBeCalledWith(values);
    expect(instance.state.updatedNote).toEqual('Yummy!');
    expect(instance.handleNoteEditor).toHaveBeenCalled();
  });

  it('can handleNoteSubmit delete', () => {
    expect(instance).toBeDefined();
    const values = { flavorId: 2, note: null, update: false };

    instance.handleNoteSubmit(values);
    expect(actions.deleteNote).toBeCalledWith(values);
    expect(instance.state.updatedNote).toEqual('');
    expect(instance.handleNoteEditor).toHaveBeenCalled();
  });
});
