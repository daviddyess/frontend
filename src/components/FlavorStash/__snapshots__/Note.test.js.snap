// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Note /> can get add note icon 1`] = `
<React.Fragment>
  <Button
    active={false}
    className="button-animation"
    disabled={false}
    onClick={[Function]}
    size="sm"
    type="button"
    variant="primary"
  >
    <FontAwesomeIcon
      icon="pen"
      size="sm"
      title="Edit Flavor Note"
    />
  </Button>
</React.Fragment>
`;

exports[`<Note /> can get note edit icon 1`] = `
<React.Fragment>
  <Button
    active={false}
    className="button-animation"
    disabled={false}
    onClick={[Function]}
    size="sm"
    type="button"
    variant="primary"
  >
    <FontAwesomeIcon
      icon="pen"
      size="sm"
      title="Edit Flavor Note"
    />
  </Button>
</React.Fragment>
`;

exports[`<Note /> can render 1`] = `
<h3>
  <button
    className="button-animation btn btn-primary btn-sm"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <i
      className="fa pen"
    />
  </button>
   
  Flavor Notes
</h3>
`;

exports[`<Note /> can render with an empty note 1`] = `
<h3>
  <button
    className="button-animation btn btn-primary btn-sm"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <i
      className="fa plus"
    />
  </button>
   
  Flavor Notes
</h3>
`;
